### PR TITLE
Add basic support for multi factor authentication

### PR DESCRIPTION
**Pull request by TimWolla**
This pull request adds basic support for multifactor authencation. The user is
able to set up authentication using emergency codes and TOTP. They are also
able to authenticate with a multifactor method within the frontend if the
`multifactorActive` flag is manually enabled within the database. This does not
yet happen automatically after setting up a multifactor method. Login within
the ACP is not yet possible. The necessary form and templates are not yet
copied over. In addition the redirect after the login goes to the frontend into
of staying within the ACP.

The UI is pretty rough as of right now, containing mostly functional buttons
without much explanation. It definitely needs a face lift once the general
workflow is approved and everything is in place.

see #3711

-------

- Add com.woltlab.wcf.multifactor objectTypeDefinition
- Add com.woltlab.wcf.multifactor.totp objectType
- Add wcf1_user_multifactor
- Integrate multifactor into AccountSecurityPage
- Add IMultifactorMethod
- Add com.woltlab.wcf.multifactor.backup object type
- Add User::getEnabledMultifactorMethods()
- Add MultifactorManageForm
- Implement getStatusText() for the `backup` multifactor method
- Add support for adding devices to TotpMultifactorMethod
- Add SessionHandler::changeUserAfterMultifactor()
- Add MultifactorAuthenticationForm
- Add QR code to __totpSecretField.tpl
- Add authentication support to TotpMultifactorMethod
- Add multifactor language items
- Face lift adding TOTP devices
- Add barebones support for deleting TOTP devices
- Add default device name for TOTP
- Add namespace to all functions and constants for BackupMultifactorMethod
- Add namespace to all functions and constants for TotpMultifactorMethod
- Add namespace to all functions and constants for multifactor forms
- Move the heavy TOTP lifting into unpack

### CODE REVIEW

File: com.woltlab.wcf/templates/__totpNewDeviceContainer.tpl [hunk 1]
```diff
@@ -0,0 +1,44 @@
+<section id="{@$container->getPrefixedId()}Container"{*
+	*}{if !$container->getClasses()|empty} class="{implode from=$container->getClasses() item='class' glue=' '}{$class}{/implode}" {/if}{*
```
Comment (Gravatronics) [inline com.woltlab.wcf/templates/__totpNewDeviceContainer.tpl:2]:
```suggestion
	*}{if !$container->getClasses()|empty} class="{implode from=$container->getClasses() item='class' glue=' '}{$class}{/implode}"{/if}{*
```

File: com.woltlab.wcf/templates/__totpNewDeviceContainer.tpl [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+		{/if}
+		
+		<div class="multifactorTotpNewDeviceFields">
+			{foreach from=$container item='child'}
+				{if $child->getId() !== 'secret' && $child->getId() !== 'submitButton' && $child->isAvailable()}
+					{@$child->getHtml()}
+				{/if}
+			{/foreach}
+			
+			<div class="formSubmit">
```
Comment (Gravatronics) [inline com.woltlab.wcf/templates/__totpNewDeviceContainer.tpl:29]: Is the `.formSubmit` element always available on purpose?

File: com.woltlab.wcf/templates/__totpSecretField.tpl [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+	<kbd {*
+	*}class="totpSecret" {*
+	*}data-issuer="{PAGE_TITLE}" {*
+	*}data-accountname="{$__wcf->user->username}"{*
+	*}>{$field->getEncodedValue()}</kbd>
+<script>
+(function (script) {
+	require(['WoltLabSuite/Core/Ui/User/Multifactor/Totp/Qr'], (Qr) => Qr.render(script.closest(".totpSecretContainer")));
+})(document.currentScript);
+</script>
```
Comment (Gravatronics) [inline com.woltlab.wcf/templates/__totpSecretField.tpl:14]: Missing indentation

**Review by Gravatronics — CHANGES_REQUESTED** [review]

File: com.woltlab.wcf/templates/__totpNewDeviceContainer.tpl (hunk 1 above)
Comment (TimWolla) [reply com.woltlab.wcf/templates/__totpNewDeviceContainer.tpl:2]: This is a common issue in all containers. I'll fix that in master.

File: com.woltlab.wcf/templates/__totpNewDeviceContainer.tpl (hunk 2 above)
Comment (Gravatronics) [reply com.woltlab.wcf/templates/__totpNewDeviceContainer.tpl:31]: No, those checks should be there.

Comment (TimWolla) [reply com.woltlab.wcf/templates/__totpNewDeviceContainer.tpl:31]: Okay. Would you recommend moving the `<div>` into the `{if $container->getNodeById('submitButton')->isAvailable()}` then?

Comment (Gravatronics) [reply com.woltlab.wcf/templates/__totpNewDeviceContainer.tpl:31]: Yes.

File: wcfsetup/install/files/lib/system/user/multifactor/totp/Totp.class.php [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+	/**
+	 * Validates the given userCode against the given minimum counter and time.
+	 * 
+	 * If this method returns `true` the $minCounter value will be updated to the counter that
+	 * was used for verification. You MUST store the updated $minCounter to prevent code re-use.
+	 */
+	public function validateTotpCode(string $userCode, int &$minCounter, \DateTime $time): bool {
+		$counter = \intval($time->getTimestamp() / self::TIME_STEP);
+		
+		for ($offset = -2; $offset < 2; $offset++) {
```
Comment (TimWolla) [inline wcfsetup/install/files/lib/system/user/multifactor/totp/Totp.class.php:71]: This 2 should move into a class constant.

File: wcfsetup/install/files/lib/system/user/multifactor/Setup.class.php [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+			FROM	wcf".WCF_N."_user_multifactor
+			WHERE	setupId = ?
+			FOR UPDATE";
+		$statement = WCF::getDB()->prepareStatement($sql);
+		$statement->execute([
+			$this->getId(),
+		]);
+		
+		$setupId = \intval($statement->fetchSingleColumn());
+		assert($setupId === $this->getId());
```
Comment (TimWolla) [inline wcfsetup/install/files/lib/system/user/multifactor/Setup.class.php:63]:
```suggestion
		\assert($setupId === $this->getId());
```

File: wcfsetup/install/files/lib/system/user/multifactor/Setup.class.php [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+		$sql = "SELECT	*
+			FROM	wcf".WCF_N."_user_multifactor
+			WHERE	userID = ?
+			FOR UPDATE";
+		$statement = WCF::getDB()->prepareStatement($sql);
+		$statement->execute([$user->userID]);
+		
+		$setups = [];
+		while ($row = $statement->fetchArray()) {
+			$setups[$row['setupID']] = new Setup($row);
```
Comment (TimWolla) [inline wcfsetup/install/files/lib/system/user/multifactor/Setup.class.php:104]:
```suggestion
			$setups[$row['setupID']] = new self($row);
```

File: wcfsetup/install/files/lib/system/user/multifactor/Setup.class.php [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+					(userID, objectTypeID)
+			VALUES		(?, ?)";
+		$statement = WCF::getDB()->prepareStatement($sql);
+		$statement->execute([
+			$user->userID,
+			$objectType->objectTypeID,
+		]);
+		
+		$setup = self::find($objectType, $user);
+		assert($setup);
```
Comment (TimWolla) [inline wcfsetup/install/files/lib/system/user/multifactor/Setup.class.php:124]:
```suggestion
		\assert($setup);
```

File: wcfsetup/install/files/lib/form/MultifactorManageForm.class.php [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+	 * Returns the Object type representing backup codes.
+	 */
+	protected function getBackupCodesObjectType(): ObjectType {
+		return ObjectTypeCache::getInstance()->getObjectTypeByName('com.woltlab.wcf.multifactor', 'com.woltlab.wcf.multifactor.backup');
+	}
+	
+	/**
+	 * Returns whether backup codes are set up yet.
+	 */
+	protected function areBackupCodesSetup(): bool {
```
Comment (dtdesign) [inline wcfsetup/install/files/lib/form/MultifactorManageForm.class.php:170]:
```suggestion
	protected function hasBackupCodes(): bool {
```
More in line with similar methods and sounds much cleaner.

**Review by dtdesign — CHANGES_REQUESTED** [review]

File: wcfsetup/install/files/lib/acp/form/LoginForm.class.php [hunk 9]
```diff
@@ -200,16 +202,20 @@ public function save() {
...
+		$this->performRedirect($needsMultifactor);
 	}
 	
 	/**
 	 * Performs the redirect after successful authentication.
 	 */
-	protected function performRedirect() {
+	protected function performRedirect(bool $needsMultifactor = false) {
+		if ($needsMultifactor) {
+			$this->url = LinkHandler::getInstance()->getControllerLink(MultifactorAuthenticationForm::class);
```
[the remark this answers is not in the document]
Comment (TimWolla) [reply wcfsetup/install/files/lib/acp/form/LoginForm.class.php:216]: I'll handle this, together with all the other (UX) issues within a follow up PR. I expect this to change / be cleaned up anyway.

**Pull request merged**